tools/wptrunner/wptrunner/testloader.py

```diff
@@ -29,29 +31,103 @@ def do_delayed_imports():
     from manifest.download import download_from_github  # type: ignore
 
 
-class TestGroupsFile:
-    """
-    Mapping object representing {group name: [test ids]}
-    """
-
-    def __init__(self, logger, path):
+class TestGroups:
+    def __init__(self, logger, path, subsuites):
         try:
             with open(path) as f:
-                self._data = json.load(f)
+                data = json.load(f)
         except ValueError:
             logger.critical("test groups file %s not valid json" % path)
             raise
 
+        self.tests_by_group = defaultdict(set)
         self.group_by_test = {}
-        for group, test_ids in self._data.items():
+        for group, test_ids in data.items():
             for test_id in test_ids:
-                self.group_by_test[test_id] = group
+                subsuite, group_name = group.split(":", 1)
```

Comment:
This is saying all tests in a group can only be from the same subsuite, I feel this is an unnecessary constraint.

Reply:
It's a pretty fundamental constraint of the design. When we queue up `TestGroup` items for execution each `TestGroup` has exactly one subsuite. For Chrome this feels like a pretty necessary limitation if we always require a browser restart for each subsuite (in order to change the pref settings). For other browser (like Firefox) we could often don't need a restart for external reasons, but it still seems like a rather reasonable choice because a) it makes the implementation significantly less complex and b) it provides isolation between different subsuites.

Reply:
Thanks, I think your explanations make sense.